freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/UserSyncService.java

```diff
@@ -358,14 +351,70 @@ private SyncStatusDetail internalSynchronizeStack(Stack stack, UmsUsersState ums
                 LOGGER.debug("Finished {}.", LogEvent.SYNC_CLOUD_IDENTITIES);
             }
 
+            retrySyncIfBatchCallHasWarnings(stack, umsUsersState, fullSync, warnings, freeIpaClient,
+                    fmsToFreeipaBatchCallEnabled, usersStateDifferenceBeforeSync);
             return toSyncStatusDetail(environmentCrn, warnings);
         } catch (Exception e) {
             LOGGER.warn("Failed to synchronize environment {}", environmentCrn, e);
             return SyncStatusDetail.fail(environmentCrn, e.getLocalizedMessage(), warnings);
         }
     }
 
-    private SyncStatusDetail internalSynchronizeStackForDeleteUser(Stack stack, String deletedWorkloadUser) {
+    private void retrySyncIfBatchCallHasWarnings(Stack stack, UmsUsersState umsUsersState, boolean fullSync, Multimap<String, String> warnings,
+            FreeIpaClient freeIpaClient, boolean fmsToFreeipaBatchCallEnabled, UsersStateDifference usersStateDifferenceBeforeSync)
+            throws FreeIpaClientException, IOException {
+        if (fullSync && !warnings.isEmpty() && fmsToFreeipaBatchCallEnabled) {
+            UsersStateDifference usersStateDifferenceAfterSync = compareUmsAndFreeIpa(umsUsersState, fullSync, freeIpaClient);
+            if (usersStateDifferenceChanged(usersStateDifferenceBeforeSync, usersStateDifferenceAfterSync)) {
+                LOGGER.info(String.format("Sync was partially successful for %s, thus we are trying it once again", stack.getResourceCrn()));
+                applyDifference(umsUsersState, stack.getEnvironmentCrn(), warnings, usersStateDifferenceAfterSync, freeIpaClient, fmsToFreeipaBatchCallEnabled);
```

Comment:
I thought a little more about the warnings. The retry is used to resolve the warnings so it doesn't make sense to return all the original warnings. What do you think of this?
```
Multimap<String, String> retryWarnings = ArrayListMultimap.create();
try {
    applyDifference(umsUsersState, stack.getEnvironmentCrn(), retryWarnings, usersStateDifferenceAfterSync, freeIpaClient, fmsToFreeipaBatchCallEnabled);
    warnings.clear();
} finally {
    warnings.putAll(retryWarnings);
}
```
The original warnings are combined with the retry warnings used if applyDifference raises an exception while only the retry warnings are used if applyDifference completes.

Reply:
i'll put this into the retry method